plugins/linux-capture/xcomposite-input.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <obs-module.h>
+#include <X11/Xlib.h>
+#include <xcb/xcb.h>
+#include <glad/glad.h>
+#include <glad/glad_glx.h>
+
+#include <string>
+#include <list>
+#include <cstdint>
+#include <cstdbool>
+
+#include "xcursor-xcb.h"
+
+struct xcompcap {
+	obs_source_t *source;
+
+	const char *windowName;
+	xcb_window_t win;
+	int crop_top;
+	int crop_left;
+	int crop_right;
+	int crop_bot;
+	bool swapRedBlue;
+	bool include_border;
+	bool exclude_alpha;
+
+	float window_check_time;
+	bool window_changed;
+
+	uint32_t width;
+	uint32_t height;
+	uint32_t border;
+
+	Pixmap pixmap;
+	GLXPixmap glxpixmap;
+	gs_texture_t *tex;
+	gs_texture_t *gltex;
+
+	pthread_mutex_t lock;
+
+	bool show_cursor;
+	bool cursor_outside;
+	xcb_xcursor_t *cursor;
+	// strict_binding determines whether we rebind the GLX Pixmap on every
+	// tick. "Strict binding" is the correct mode of operation, according to
+	// GLX_EXT_texture_from_pixmap. However certain drivers exhibit poor
+	// performance when this is done, so setting this to false allows working
+	// around it.
+	bool strict_binding;
+};
+
+// Defer implementation from http://the-witness.net/news/2012/11/scopeexit-in-c11/
+template<typename F> struct ScopeExit {
+	ScopeExit(F f) : f(f) {}
+	~ScopeExit() { f(); }
+	F f;
+};
+
+template<typename F> ScopeExit<F> MakeScopeExit(F f)
+{
+	return ScopeExit<F>(f);
+};
+
+#define XSTRING_JOIN2(arg1, arg2) STRING_JOIN2(arg1, arg2)
+#define STRING_JOIN2(arg1, arg2) arg1##arg2
+#define DEFER(code)                                 \
+	auto XSTRING_JOIN2(scope_exit_, __LINE__) = \
+		MakeScopeExit([=]() { code; })
```

Comment:
Is this the only reason for using C++ here? If so, allow me to suggest `__attribute__((__cleanup__(cleanup_func)))`, which can be similarly wrapped in a macro like this:

```c
#define autocleanup(cleanup_func) __attribute__((__cleanup__(cleanup_func)))
```

You can then use it in the code as follows:

```c
autocleanup(free) char *string_to_be_freed_by_the_end_of_the_scope = NULL;
autocleanup(XFree) GLXFBConfig config = NULL;
```

It's important to initialize these values as NULL otherwise exiting the scope could lead to freeing uninitialized memory.

If this is enough to cover all C++ features used here, it would be good to make these files .c and .h after :slightly_smiling_face:

Reply:
No, I left c++ for xcompcap_props/xcomp_top_level_windows/xcomp_find_window support for some string handling that would grow a fair bit without std::string and stdlib's lexicographical sort which didnt seem to have a libc equivalent so I left them c++.

Since i kept c++ I left some usage of this little defer where it was nice to have, but its not a reason to keep c++.

Reply:
Alright, that's reasonable. My suggestion above still stands though - I feel like the autocleanups are easier to reason about compared to the DEFER() approach

Reply:
This comment still applies, if you want scoped variables - or, alternatively, you could just drop the autocleanup changes entirely (though I personally enjoy scoped variables with the `autocleanup` approach above)